tools/build/android/README.md

```diff
@@ -0,0 +1,36 @@
+# Build MixedReality-WebRTC for Android
+
+## Prerequisites
+
+1. Build the Chromium WebRTC base library. Instructions [here](../libwebrtc/README.md).
+
+2. Install prerequisites:
+    * Android Studio (https://developer.android.com/studio/index.html). Once installed, use the System Settings/Android SDK dialog to install the following packages:
+        * SDKs 9.0 (Pie), 8.1 (Oreo)
+        * CMake
+        * NDK 16.1.4479499
```

Comment:
There's no option to select a version of the NDK as I know. I can only see "NDK (side by side)", which as of today resolves to 21.0.6113669. I _think_ that's OK but I can't remember the details about NDK versions.

Reply:
Actually that's not even enough, I had to uncheck "Hide obsolete packages", then install "NDK (Obsolete)" version 21.0.6113669, then click "Sync Project with Gradle Files" before I was finally allowed to Build > Make Project.

Reply:
This section is light on detail, due entirely to the fact that I wasn't sure what the best practice is here. I wanted to provide steps that could be done entirely on the command line and not rely on Android Studio. I haven't gone back and worked out what those steps are yet.

Reply:
We can polish later. I want to commit to `master` an experimental support version, then we refine the instructions.

Reply:
Trying to build in Android Studio and getting this build error.  any suggestions?

Build command failed.
Error while executing process /home/astaikos/Android/Sdk/cmake/3.10.2.4988404/bin/cmake with arguments {--build /home/astaikos/eanders/MixedReality-WebRTC/tools/build/android/webrtc-native/.externalNativeBuild/cmake/debug/arm64-v8a --target mrwebrtc}
[1/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/interop/local_video_track_interop.cpp.o
[2/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/data_channel.cpp.o
[3/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/str.cpp.o
[4/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/local_video_track.cpp.o
[5/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/mrs_errors.cpp.o
[6/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/audio_frame_observer.cpp.o
[7/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/interop/external_video_track_source_interop.cpp.o
[8/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/interop/peer_connection_interop.cpp.o
[9/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/pch.cpp.o
[10/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/external_video_track_source.cpp.o
[11/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/interop/global_factory.cpp.o
[12/17] Building CXX object CMakeFiles/mrwebrtc.dir/jni_onload.cpp.o
[13/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/sdp_utils.cpp.o
[14/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/interop/interop_api.cpp.o
[15/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp.o
[16/17] Building CXX object CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/video_frame_observer.cpp.o
[17/17] Linking CXX shared library /home/astaikos/eanders/MixedReality-WebRTC/tools/build/android/webrtc-native/build/intermediates/cmake/debug/obj/arm64-v8a/libmrwebrtc.so
FAILED: /home/astaikos/eanders/MixedReality-WebRTC/tools/build/android/webrtc-native/build/intermediates/cmake/debug/obj/arm64-v8a/libmrwebrtc.so 
: && /home/astaikos/Android/Sdk/ndk/21.0.6113669/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=aarch64-none-linux-android24 --gcc-toolchain=/home/astaikos/Android/Sdk/ndk/21.0.6113669/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/home/astaikos/Android/Sdk/ndk/21.0.6113669/toolchains/llvm/prebuilt/linux-x86_64/sysroot -fPIC -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security  -std=c++17 -DANDROID_STL=c++_shared -DMRS_USE_STR_WRAPPER -DMR_SHARING_ANDROID -DWEBRTC_POSIX -DWEBRTC_ANDROID -O0 -fno-limit-debug-info  -Wl,--exclude-libs,libgcc_real.a -Wl,--exclude-libs,libatomic.a -static-libstdc++ -Wl,--build-id -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments -shared -Wl,-soname,libmrwebrtc.so -o /home/astaikos/eanders/MixedReality-WebRTC/tools/build/android/webrtc-native/build/intermediates/cmake/debug/obj/arm64-v8a/libmrwebrtc.so CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/interop/external_video_track_source_interop.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/interop/global_factory.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/interop/local_video_track_interop.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/interop/peer_connection_interop.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/interop/interop_api.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/external_video_track_source.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/local_video_track.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/audio_frame_observer.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/data_channel.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/mrs_errors.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/pch.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/sdp_utils.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/str.cpp.o CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/video_frame_observer.cpp.o CMakeFiles/mrwebrtc.dir/jni_onload.cpp.o  -llog -landroid -ldl -lOpenSLES /home/astaikos/eanders/build_webrtc/webrtc/src/out/android/arm64/Release/obj/libwebrtc.a -latomic -lm && :
CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp.o: In function `PeerConnection':
/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp:1370: undefined reference to `Microsoft::MixedReality::WebRTC::TrackedObject::TrackedObject(Microsoft::MixedReality::WebRTC::RefPtr<Microsoft::MixedReality::WebRTC::GlobalFactory>, Microsoft::MixedReality::WebRTC::ObjectType)'
CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp.o: In function `~PeerConnection':
/home/astaikos/eanders/MixedReality-WebRTC/tools/build/android/webrtc-native/src/main/cpp/../../../../../../../libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.h:65: undefined reference to `Microsoft::MixedReality::WebRTC::TrackedObject::~TrackedObject()'
CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp.o:(.data.rel.ro._ZTIN9Microsoft12MixedReality6WebRTC14PeerConnectionE[_ZTIN9Microsoft12MixedReality6WebRTC14PeerConnectionE]+0x10): undefined reference to `typeinfo for Microsoft::MixedReality::WebRTC::TrackedObject'
CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/external_video_track_source.cpp.o: In function `ExternalVideoTrackSource':
/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/external_video_track_source.cpp:287: undefined reference to `Microsoft::MixedReality::WebRTC::TrackedObject::TrackedObject(Microsoft::MixedReality::WebRTC::RefPtr<Microsoft::MixedReality::WebRTC::GlobalFactory>, Microsoft::MixedReality::WebRTC::ObjectType)'
CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/external_video_track_source.cpp.o: In function `~ExternalVideoTrackSource':
/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/external_video_track_source.h:79: undefined reference to `Microsoft::MixedReality::WebRTC::TrackedObject::~TrackedObject()'
CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/external_video_track_source.cpp.o:(.data.rel.ro._ZTIN9Microsoft12MixedReality6WebRTC24ExternalVideoTrackSourceE[_ZTIN9Microsoft12MixedReality6WebRTC24ExternalVideoTrackSourceE]+0x10): undefined reference to `typeinfo for Microsoft::MixedReality::WebRTC::TrackedObject'
CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/local_video_track.cpp.o: In function `LocalVideoTrack':
/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/local_video_track.cpp:18: undefined reference to `Microsoft::MixedReality::WebRTC::TrackedObject::TrackedObject(Microsoft::MixedReality::WebRTC::RefPtr<Microsoft::MixedReality::WebRTC::GlobalFactory>, Microsoft::MixedReality::WebRTC::ObjectType)'
/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/local_video_track.cpp:27: undefined reference to `Microsoft::MixedReality::WebRTC::TrackedObject::~TrackedObject()'
CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/local_video_track.cpp.o: In function `~LocalVideoTrack':
/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/local_video_track.cpp:35: undefined reference to `Microsoft::MixedReality::WebRTC::TrackedObject::~TrackedObject()'
/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/local_video_track.cpp:35: undefined reference to `Microsoft::MixedReality::WebRTC::TrackedObject::~TrackedObject()'
CMakeFiles/mrwebrtc.dir/home/astaikos/eanders/MixedReality-WebRTC/libs/Microsoft.MixedReality.WebRTC.Native/src/media/local_video_track.cpp.o:(.data.rel.ro._ZTIN9Microsoft12MixedReality6WebRTC15LocalVideoTrackE+0x28): undefined reference to `typeinfo for Microsoft::MixedReality::WebRTC::TrackedObject'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.

Reply:
I fixed that like 5 min ago, this was due to the merge of `master`, you're too fast... 😉

Reply:
I'm also attempting to build for android, via android studio. but I'm coming across the following error:
In file included from C:\MixedRealityRTCLib\MixedReality-WebRTC\libs\Microsoft.MixedReality.WebRTC.Native\src\media\external_video_track_source.cpp:4:
C:\MixedRealityRTCLib\MixedReality-WebRTC\libs\Microsoft.MixedReality.WebRTC.Native\src\pch.h:44:10: fatal error: 'api/audio_codecs/builtin_audio_decoder_factory.h' file not found

is there some include directive that is missing? I've pulled from the 'master'  branch.